pkg/reconciler/taskrun/resources/validate_params.go

```diff
@@ -0,0 +1,51 @@
+package resources
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"github.com/tektoncd/pipeline/pkg/substitution"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// ValidateParamArrayIndex validates if the param reference to an array param is out of bound.
+// error is returned when the array indexing reference is out of bound of the array param
+// e.g. if a param reference of $(params.array-param[2]) and the array param is of length 2.
+// - `trParams` are params from taskrun.
+// - `taskSpec` contains params declarations.
```

Comment:
Sorry I probably forgot to update this, do you mind updating it?

Reply:
done. I also moved some duplicated logic from pipelinerun/resources to taskrun/resources.